### PR TITLE
Add Memory String Lock provider

### PR DESCRIPTION
**Pull request by FireDefend**
Add Key Lock interface
Add Memory Key Lock implementation

User can acquire lock at unique string level

### CODE REVIEW

File: coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	return func() {
+		mslp.lm.getLockNode(key).Unlock()
+		go mslp.lm.updateLockLRU(key)
+	}
+}
+
+type LockNode struct {
+	lastUsedTime time.Time
+	key          string
+	lock         sync.Mutex
```
Comment (LeiQL) [inline coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go:81]: Why maintain the lock as linked list? Seems GET lock is not efficient.

Comment (msftcoderdjw) [reply coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go:81]: I share the same concerns with Lency. I can understand you want to keep LRU order for purging. How about we keep a map for efficient GET?

Comment (FireDefend) [reply coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go:81]: getlock will using a concurrent map 

```
locknode, _ := lm.m.LoadOrStore(key, &LockNode{key: key, prev: nil, next: nil})
```

**Review by RemindD — APPROVED** [review]

File: coa/pkg/apis/v1alpha2/providers/stringlock/stringlock.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import (
+	providers "github.com/eclipse-symphony/symphony/coa/pkg/apis/v1alpha2/providers"
+	//"encoding/json"
+)
+
+type UnLock func()
+
+type IStringLockProvider interface {
+	Init(config providers.IProviderConfig) error
+	Lock(string) UnLock
```
Comment (RemindD) [inline coa/pkg/apis/v1alpha2/providers/stringlock/stringlock.go:18]: This interface looks strange to me. Usually Lock function should return an error indicating whether lock is successful or not. And another Unlock function in the interface to unlock a key.

**Review by RemindD — COMMENTED** [review]
Mis click the approve previously. Please see my below comments.

File: coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	prev         *LockNode
+	next         *LockNode
+}
+
+type LockManager struct {
```
[the remark this answers is not in the document]
Comment (FireDefend) [reply coa/pkg/apis/v1alpha2/providers/stringlock/memorystringlock/memorystringlock.go]: changed, thanks for reminding

File: coa/pkg/apis/v1alpha2/providers/stringlock/stringlock.go (hunk 2 above)
Comment (FireDefend) [reply coa/pkg/apis/v1alpha2/providers/stringlock/stringlock.go]: added "trylock" and "unlock" interface

type IKeyLockProvider interface {
	Init(config providers.IProviderConfig) error
	Lock(string)
	UnLock(string)
	TryLock(string) bool
	TryLockWithTimeout(string, time.Duration) bool
}

**Review by msftcoderdjw — DISMISSED** [review]

**Review by msftcoderdjw — APPROVED** [review]

File: coa/pkg/apis/v1alpha2/providers/keylock/memory/memory.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		return false
+	}
+
+	node := lm.tail.prev
+	if node.prev != nil {
+		node.prev.next = node.next
+	}
+	if node.next != nil {
+		node.next.prev = node.prev
+	}
```
Comment (RemindD) [inline coa/pkg/apis/v1alpha2/providers/keylock/memory/memory.go:172]: node.next should always be lm.tail?

Comment (FireDefend) [reply coa/pkg/apis/v1alpha2/providers/keylock/memory/memory.go:172]: yes, the condition can be deleted here after I used a dummy head and tail. Will delete those "if", in the next fix pr

Comment (RemindD) [reply coa/pkg/apis/v1alpha2/providers/keylock/memory/memory.go:172]: sounds good

**Pull request merged**